sc2/observer_ai.py

```diff
@@ -0,0 +1,383 @@
+    ) -> List[List[AbilityId]]:
+        """ Returns available abilities of one or more units. Right now only checks cooldown, energy cost, and whether the ability has been researched.
+
+        Examples::
+
+            units_abilities = await self.get_available_abilities(self.units)
+
+        or::
+
+            units_abilities = await self.get_available_abilities([self.units.random])
+
+        :param units:
+        :param ignore_resource_requirements: """
+        return await self._client.query_available_abilities(units, ignore_resource_requirements)
+
+    @property_cache_once_per_frame
+    def _abilities_all_units(self) -> Counter:
+        """ Cache for the already_pending function, includes protoss units warping in,
+        all units in production and all structures, and all morphs """
+        abilities_amount = Counter()
+        for unit in self.units + self.structures:  # type: Unit
+            for order in unit.orders:
+                abilities_amount[order.ability] += 1
+            if not unit.is_ready:
+                if self.race != Race.Terran or not unit.is_structure:
+                    # If an SCV is constructing a building, already_pending would count this structure twice
+                    # (once from the SCV order, and once from "not structure.is_ready")
+                    abilities_amount[self._game_data.units[unit.type_id.value].creation_ability] += 1
+
+        return abilities_amount
+
+    def _prepare_start(self, client, player_id, game_info, game_data, realtime: bool = False):
+        """
+        Ran until game start to set game and player data.
+
+        :param client:
+        :param player_id:
+        :param game_info:
+        :param game_data:
+        :param realtime:
+        """
+        self._client: Client = client
+        self.player_id: int = player_id
+        self._game_info: GameInfo = game_info
+        self._game_data: GameData = game_data
+        self.realtime: bool = realtime
+
+    def _prepare_first_step(self):
+        """First step extra preparations. Must not be called before _prepare_step."""
+        if self.townhalls:
+            self._game_info.player_start_location = self.townhalls.first.position
+        self._game_info.map_ramps, self._game_info.vision_blockers = self._game_info._find_ramps_and_vision_blockers()
+
+    def _prepare_step(self, state, proto_game_info):
+        """
+        :param state:
+        :param proto_game_info:
+        """
+        # Set attributes from new state before on_step."""
+        self.state: GameState = state  # See game_state.py
+        # Required for events, needs to be before self.units are initialized so the old units are stored
+        self._units_previous_map: Dict = {
+            unit.tag: unit for unit in self.units}
+        self._structures_previous_map: Dict = {
+            structure.tag: structure for structure in self.structures}
+
+        self._prepare_units()
+
+    def _prepare_units(self):
```

Comment:
All units combined into self.units. This was done because who is the "enemy" if you are observing both players? Some thought required here